### PR TITLE
Add tracing context to storage blobs

### PR DESCRIPTION
**Pull request by samvaity**
- Added tracing context to sync and async storage blob clients.
- Added T with Response<T> "overload in sync and async storage blob clients.

**NOTE**
Left the doc updates be as they would be anyway updated by @alzimmermsft later.
Flux return types to be updated in later PR's

Closes: #4329

### CODE REVIEW

File: storage/client/blob/src/test/java/com/azure/storage/blob/Sample.java [hunk 1]
```diff
@@ -80,74 +80,74 @@ public void sample() throws IOException {
...
-                    .map(buffer -> new String(buffer.array()))
-                    .doOnNext(string -> System.out.println(listItem.name() + ": " + string)))
-            // cleanup
-            .thenMany(serviceClient.listContainers())
-            .flatMap(containerItem -> serviceClient
-                .getContainerAsyncClient(containerItem.name())
-                .delete())
-            .blockLast();
-    }
+//    public void asyncSample() throws IOException {
```
Comment (alzimmermsft) [inline storage/client/blob/src/test/java/com/azure/storage/blob/Sample.java]: Why did all of this get commented out?

File: storage/client/blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java [hunk 2]
```diff
@@ -69,10 +71,10 @@
      * Creates a 0-length append blob. Call appendBlock to append data to an append blob.
      *
      * @return
-     *      A reactive response containing the information of the created appended blob.
+     *      A {@link Mono} containing the information of the created appended blob.
```
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: I had seen this pattern of docs in keyvault not sure if we follow it in other clients as well. Do you suggest not changing these?

**Comment by samvaity** [comment]
@alzimmermsft All the javadoc changes as mentioned above is still a TODO.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java [hunk 3]
```diff
@@ -65,9 +66,9 @@ public ContainerClient getContainerClient(String containerName) {
      * <a href="https://docs.microsoft.com/rest/api/storageservices/create-container">Azure Docs</a>.
      *
      * @param containerName Name of the container to create
-     * @return A response containing a {@link ContainerClient} used to interact with the container created.
+     * @return The {@link ContainerClient} used to interact with the container created.
      */
-    public Response<ContainerClient> createContainer(String containerName) {
+    public ContainerClient createContainer(String containerName) {
```
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java]: We do have a Response type equivalent of this method. Will that not work for the above concern?

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 4]
```diff
@@ -165,13 +166,30 @@ public URL getBlobUrl() {
...
+     *
+     * @return true if the blob exists, false if it doesn't
+     */
+    public Mono<Response<Boolean>> existsWithResponse() {
+        return withContext(context -> existsWithResponse(context));
+    }
+
+    Mono<Response<Boolean>> existsWithResponse(Context context) {
+        return this.getPropertiesWithResponse(null, context)
+                   .map(cp -> (Response<Boolean>) new SimpleResponse<>(cp, true))
```
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: Didn't get any checkstyle issues on these indentations...

File: storage/client/blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: left the doc updates as those would be covered later on.

File: storage/client/blob/src/main/java/com/azure/storage/blob/ContainerClient.java [hunk 5]
```diff
@@ -297,7 +322,7 @@ public VoidResponse delete(ContainerAccessConditions accessConditions, Duration
      * @return A response containing status code and HTTP headers
      */
     public VoidResponse setMetadata(Metadata metadata) {
-        return this.setMetadata(metadata, null, null);
+        return this.setMetadata(metadata, null, null, Context.NONE);
```
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/ContainerClient.java]: not using this.

File: storage/client/blob/src/test/java/com/azure/storage/blob/Sample.java (hunk 1 above)
Comment (samvaity) [reply storage/client/blob/src/test/java/com/azure/storage/blob/Sample.java]: uncommented

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java]: It should work, there might be a little more effort on the consumer side to know that getX is code only while createX makes a service call.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java [hunk 6]
```diff
@@ -50,12 +52,12 @@
     /**
      * Indicates the maximum number of bytes that can be sent in a call to appendBlock.
      */
-    public static final int MAX_APPEND_BLOCK_BYTES = 4 * Constants.MB;
+    public final static int MAX_APPEND_BLOCK_BYTES = 4 * Constants.MB;
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: Did the Checkstyle rule change to order `final static` instead of `static final` as it has been?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java [hunk 7]
```diff
@@ -86,16 +86,20 @@
      *         {@link BlobAccessConditions}
      *
      * @return
-     *      A reactive response containing the information of the created appended blob.
+     *      A {@link Mono} containing {@link Response} whose {@link Response#value() value} contains the {created appended blob.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: Random `{` that'll break JavaDoc generation

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (samvaity) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java]: I think in general other Java clients too follow the same concept for getX and createX so it should be fine.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java (hunk 6 above)
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: nope, fixed!

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java (hunk 7 above)
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]: yup, removed!

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/PageBlobAsyncClient.java [hunk 8]
```diff
@@ -110,8 +112,13 @@
      * @throws IllegalArgumentException If {@code size} isn't a multiple of {@link PageBlobAsyncClient#PAGE_BYTES}
      * or {@code sequenceNumber} isn't null and is less than 0.
      */
-    public Mono<Response<PageBlobItem>> create(long size, Long sequenceNumber, BlobHTTPHeaders headers,
+    public Mono<Response<PageBlobItem>> createWithResponse(long size, Long sequenceNumber, BlobHTTPHeaders headers,
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/PageBlobAsyncClient.java]: >createWithResponse [](start = 40, length = 18)

Should this be named create like the other public facing methods?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceClientBuilder.java [hunk 9]
```diff
@@ -109,8 +110,11 @@ public BlobServiceAsyncClient buildAsyncClient() {
         }
 
         policies.add(new RequestRetryPolicy(retryOptions));
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
 
         policies.addAll(this.policies);
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceClientBuilder.java]: > Shouldn't everything fall into the bucket of before or after the retry policy?

All the default policies that the library has to add can fall in this bucket.
But for any additional policies that the customer wants to set will be added using that line.

>Can we rename this.policies to this.customerPolicies? This method is hard to follow with two different lists called "policies"

Renamed it to `additionalPolicies`

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java [hunk 10]
```diff
@@ -105,27 +109,25 @@
...
      *         (the default). In other words, the Flux must produce the same data each time it is subscribed to.
      * @param length
-     *         The exact length of the data. It is important that this value match precisely the length of the data
-     *         emitted by the {@code Flux}.
+     *          The exact length of the data. It is important that this value match precisely the length of the data
+     *          emitted by the {@code Flux}.
      *
      * @return
-     *      A reactive response containing the information of the append blob operation.
+     *      {@link Mono} containing a the information of the append blob operation.
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]:
```suggestion
     *      {@link Mono} containing the information of the append blob operation.
```

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java [hunk 11]
```diff
@@ -164,16 +170,14 @@
      *          The source {@link BlobRange} to copy.
      *
      * @return
-     *      A reactive response containing the information of the append blob operation.
+     *      {@link Mono} containing a the information of the append blob operation.
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobAsyncClient.java]:
```suggestion
     *      {@link Mono} containing the information of the append blob operation.
```

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 12]
```diff
@@ -231,23 +249,27 @@ public URL getBlobUrl() {
...
         // We want to hide the SourceAccessConditions type from the user for consistency's sake, so we convert here.
         SourceModifiedAccessConditions sourceConditions = new SourceModifiedAccessConditions()
-            .sourceIfModifiedSince(sourceModifiedAccessConditions.ifModifiedSince())
-            .sourceIfUnmodifiedSince(sourceModifiedAccessConditions.ifUnmodifiedSince())
-            .sourceIfMatch(sourceModifiedAccessConditions.ifMatch())
-            .sourceIfNoneMatch(sourceModifiedAccessConditions.ifNoneMatch());
+                                                              .sourceIfModifiedSince(sourceModifiedAccessConditions.ifModifiedSince())
+                                                              .sourceIfUnmodifiedSince(sourceModifiedAccessConditions.ifUnmodifiedSince())
+                                                              .sourceIfMatch(sourceModifiedAccessConditions.ifMatch())
+                                                              .sourceIfNoneMatch(sourceModifiedAccessConditions.ifNoneMatch());
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: This kind of horizontal shuffling should be reverted.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 13]
```diff
@@ -797,10 +865,30 @@ private void downloadToFileCleanup(AsynchronousFileChannel channel) {
...
+     * {@codesnippet com.azure.storage.blob.BlobAsyncClient.undeleteWithResponse}
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/undelete-blob">Azure Docs</a></p>
+     *
+     * @return A reactive response signalling completion.
+     */
+    public Mono<VoidResponse> undeleteWithResponse() {
+        return withContext(context -> undeleteWithResponse(context));
+    }
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: I wonder if there is much value in the 'withResponse' variant when the return type of the non-withResponse variant is Mono<Void>. I wonder if in the case of Mono<Void> return types, we just change it to Mono<VoidResponse> and get rid of the 'withResponse' variety. I would worry a bit about discoverability and inconsistency, but on the upside, this lowers the method count. Thoughts?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java [hunk 14]
```diff
@@ -629,8 +639,8 @@ public VoidResponse setMetadata(Metadata metadata, BlobAccessConditions accessCo
      * @param tier The new tier for the blob.
      * @return A response containing status code and HTTP headers.
      */
-    public VoidResponse setTier(AccessTier tier) {
-        return this.setTier(tier, null, null);
+    public Void setTier(AccessTier tier) {
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: This should be `void`, correct?

Comment (JonathanGiles) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: If so, please apply in all applicable cases.

Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: To make this `void` we will have to call the `async` equivalent of this method and not the overload as we have been doing everywhere else.
Something like this,
```java
public void setTier(AccessTier tier) {
        blobAsyncClient.setTierWithResponse(tier, null, Context.NONE).block();
        return;
    }
```
still do it ?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java (hunk 13 above)
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: I think in addition to consistency, those `withResponse` return type methods could be useful to the users to know the status of the operation.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobClient.java [hunk 15]
```diff
@@ -220,10 +244,41 @@ public BlobOutputStream getBlobOutputStream(AppendBlobAccessConditions accessCon
...
+
+    /**
+     * Commits a new block of data from another blob to the end of this append blob.
+     *
+     * @param sourceURL
+     *          The url to the blob that will be the source of the copy.  A source blob in the same storage account can
+     *          be authenticated via Shared Key. However, if the source is a blob in another account, the source blob
+     *          must either be public or must be authenticated via a shared access signature. If the source blob is
+     *          public, no authentication is required to perform the operation.
+     * @param sourceRange
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/AppendBlobClient.java]: I think the javadoc is best ignored at the moment. @alzimmermsft would be soon updating these in later PR's .

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java (hunk 14 above)
Comment (JonathanGiles) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: Could you just not return anything with the code as it is today?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java (hunk 13 above)
Comment (JonathanGiles) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: Sorry, GitHub stripped out the generic typing in my previous comment. I've fixed it - please re-read it to see what I actually meant to say :-) The Response information is still present in my suggestion - but I do totally accept it is not consistent, hence why I'm seeking your opinion :-)

Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: yeah, I think I agree with this. I don't think `Mono<Void>` apart from consistency add any real benefits to the user.
Should I make this change as well?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java (hunk 14 above)
Comment (samvaity) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: updated.
